Feature/add remax support

liziniu opened the pull request:
## Description
Added [ReMax](https://arxiv.org/abs/2310.10505) support to verl.  ReMax is a simple, efficient, and stable RL algorithm customized for LLM training, with theoretical guarantees for variance reduction.

The [HybridFlow](https://arxiv.org/pdf/2409.19256v2) paper experimented with ReMax, but verl did not provide an implementation. Therefore, ReMax has been added.


## Changes
- Added RayReMaxTrainer implementation
- Added example scripts for ReMax training
- Added documentation for ReMax algorithm

## Testing
- Tested ReMax example scripts with Qwen models

validation reward of optimizing Qwen2.5-3B-Instruct on the GSM8K dataset:

<img width="501" alt="截屏2025-02-09 20 51 14" src="https://github.com/user-attachments/assets/742c2eab-6877-4c3c-b0a2-4159bd109add" />

The curve demonstrates the effectiveness of ReMax, though its performance can be further enhanced through hyperparameter fine-tuning.

## Documentation
- Added ReMax documentation
- Updated example configurations

## Checklist
- [x] Code follows project's style guidelines (yapf formatted)
- [x] Tests added/updated and passing
- [x] Documentation updated
- [x] Example scripts added

vermouth1992 commented:
Hi,

@liziniu Thank you for your contribution. According to our implementation, I guess ReMax can be implemented by adding a few lines to the original PPO/GRPO/Reinforce implementation instead of writing a new trainer to make maintenance easier. Correct me if this is invalid

PeterSH6 commented:
+1. From my understanding, Remax can be implemented similarly to reinforce++ with a different adv estimator. See reinforce++ implementation: https://github.com/volcengine/verl/pull/228

eric-haibin-lin commented on README.md:
```diff
@@ -41,7 +41,7 @@ verl is fast with:
 - huggingface models support
 - Supervised fine-tuning
-- Reinforcement learning from human feedback with [PPO](https://github.com/volcengine/verl/tree/main/examples/ppo_trainer) and [GRPO](https://github.com/volcengine/verl/tree/main/examples/grpo_trainer)
+- Reinforcement learning from human feedback with [PPO](https://github.com/volcengine/verl/tree/main/examples/ppo_trainer), [GRPO](https://github.com/volcengine/verl/tree/main/examples/grpo_trainer), and [ReMax](https://github.com/volcengine/verl/tree/main/examples/remax_trainer)
```
if you have the training log and wandb already, would you mind adding one more record to `docs/experiment/ppo.rst` to include remax? it would help the community to track if experiment can be reproduced in future version. 
We can do that in the next PR

liziniu replied:
Yes. A preliminary result on Qwen2.5-3B is added and more results will come later.

liziniu commented:
@vermouth1992 @PeterSH6 I see. Let me reformat the code with minimal changes of the PPO's trainer.

liziniu commented:
Hi, @vermouth1992 @PeterSH6 

I have completed the implementation of ReMaX support. The changes include:

1. Remove the new trainer for ReMax
2. Implemented Remax based on PPO's trainer 
3. Updated the preliminary result in `docs/experiment/ppo.rst`

The code follows the project's style guidelines.

Please review when you have a chance. Let me know if any changes or clarifications are needed.

Thank you for your time!

PeterSH6 commented on verl/trainer/ppo/core_algos.py:
```diff
@@ -154,14 +154,16 @@ def compute_grpo_outcome_advantage(token_level_rewards: torch.Tensor,
     return scores, scores
 
 
-def compute_reinforce_plus_plus_outcome_advantage(token_level_rewards: torch.Tensor, eos_mask: torch.Tensor,
```
Hi @liziniu , it seems that you accidentally remove the reinforce++ adv.
You can add a separate adv function for remax.

liziniu replied:
Thank you for pointing out this mistake. I have fixed this issue.

liziniu replied:
Thank you for pointing out this mistake. I have fixed it.

vermouth1992 commented:
Could you add a CI to run remax with Qwen 0.5b to protect this functionality?

liziniu commented:
Hi @vermouth1992 

I’ve updated the end-to-end (e2e) test. Since this is my first time writing such a test within the workflows, I’m not entirely confident about it. Could you please review it when you have a moment?

Thank you!
